### PR TITLE
[SPARK-20538][SQL] Wrap Dataset.reduce with withNewRddExecutionId.

### PR DESCRIPTION
**Pull request by sohama4**
## What changes were proposed in this pull request?

Wrap Dataset.reduce with `withNewExecutionId`.

### CODE REVIEW

**Comment by sohama4** [comment]
@vanzin @maropu @gatorsmile @jaceklaskowski can someone take a look at this? I'd appreciate it!

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -1607,7 +1607,9 @@ class Dataset[T] private[sql](
    */
   @Experimental
   @InterfaceStability.Evolving
-  def reduce(func: (T, T) => T): T = rdd.reduce(func)
+  def reduce(func: (T, T) => T): T = withNewExecutionId {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Only a place to need this wrapper in this file?

@gatorsmile btw, this api is still useful? Or, this intends to be deprecated in v3.0+?

Comment (rxin) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: cc @zsxwing

**Review by jaceklaskowski — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (jaceklaskowski) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @maropu When you asked about this API did you refer to `reduce` or `withNewExecutionId`?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: `reduce`

Comment (rxin) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Why would we want to deprecate it?

**Comment by sohama4** [comment]
Thanks for the approval @jaceklaskowski! Can you leave a comment so that Jenkins can get testing underway?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: this method should use `withNewRDDExecutionId`

**Review by zsxwing — CHANGES_REQUESTED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (sohama4) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Thanks, that makes sense when I looked at the code for `foreach` and `foreachPartition`; I put up a new version with this change. It however wasn't clear immediately how the new function `withNewRDDExecutionId` would be beneficial over `withNewExecutionId`, can you elaborate a little when you get the chance?

**Comment by kiszk** [comment]
Is it better to update title and the PR description, too?

**Comment by sohama4** [comment]
@kiszk @jaceklaskowski @rxin @zsxwing @maropu can we move forward here? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (zsxwing) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @sohama4 Spark Web UI will show no SQL metrics when using `withNewExecutionId`. See https://github.com/apache/spark/pull/19689

**Comment by zsxwing** [comment]
ok to test

**Comment by zsxwing** [comment]
LGTM pending tests

**Comment by zsxwing** [comment]
Thanks! Merging to master.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: @rxin Sorry, I missed that comment. I just thought what `reduce` can do is what type-safe `groupByKey` can do. But, probably I missed something.

**Pull request closed without merging**